### PR TITLE
feat: add promise support for client operations

### PR DESCRIPTION
**Pull request by shimks**
#### Related issues

- connect to <link_to_referenced_issue>

### Checklist

- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

File: src/client.js [hunk 1]
```diff
@@ -87,11 +88,28 @@ class Client extends Base {
...
+          callback = function(err, result, raw, soapHeader) {
+            if (err) {
+              reject(err);
+            }
+            else {
+              resolve({result, raw, soapHeader});
+            }
+          }
+        });
+        callback.promise = promise;
```
Comment (raymondfeng) [inline src/client.js]: Can you extract it into a function such as `createPromiseCallback`?

File: src/client.js [hunk 2]
```diff
@@ -87,11 +90,18 @@ class Client extends Base {
         callback = extraHeaders;
         extraHeaders = options;
         options = temp;
+      } else if (typeof callback === 'object') {
+        extraHeaders = options;
+        options = callback;
+        callback = undefined;
+      }
+      if (!callback) {
```
Comment (jannyHou) [inline src/client.js]: nitpick feel free to ignore: I think the code below can be written in a conciser way: 

```js
callback = callback || utils.createPromiseCallback();
self._invode(operation, args, location, callback, options, extraHeaders);
return callback.promise;
```

File: README.md [hunk 3]
```diff
@@ -172,6 +172,17 @@ Call *method* on the SOAP service.
       // soapHeader is the response soap header as a JavaScript object
   })
 ```
+
+A *method* can also be called as a promise.
+
+``` javascript
+  client.MyFunction({name: 'value'}).then(function({result, raw, soapHeader}){
```
Comment (bajtos) [inline README.md]: Can you add an example showing `await` usage too please? In my experience, many developers don't realize it's possible to `await` any function that's returning a promise.

```js
const {result, raw, soapHeader} = client.MyFunction({name: 'value'});
```

File: src/client.js [hunk 4]
```diff
@@ -13,7 +13,9 @@ var HttpClient = require('./http'),
   _ = require('lodash'),
   debug = require('debug')('strong-soap:client'),
   debugDetail = require('debug')('strong-soap:client:detail'),
-  debugSensitive = require('debug')('strong-soap:client:sensitive');
+  debugSensitive = require('debug')('strong-soap:client:sensitive'),
+  Promise = require('bluebird'),
```
Comment (bajtos) [inline src/client.js]: Can we use native promises and drop dependency on bluebird please?

File: src/client.js (hunk 2 above)
Comment (bajtos) [reply src/client.js]: +1

File: src/utils.js [hunk 5]
```diff
@@ -64,3 +64,19 @@ exports.toXMLDate = function(d) {
...
 };
 
+exports.createPromiseCallback = function createPromiseCallback() {
+  var cb;
+  var promise = new Promise(function(resolve, reject) {
+    cb = function(err, result, envelope, soapHeader) {
+      if (err) {
+        reject(err);
+      }
+      else {
```
Comment (bajtos) [inline src/utils.js]:
```js
} else {
```

**Comment by shimks** [comment]
Seems like `microservicex.net/stockquote.wsdl` still can't be reached. Does anyone know whether that endpoint's been discontinued or not?

**Comment by shimks** [comment]
@slnode test please

**Comment by shimks** [comment]
@slnode test please

File: README.md [hunk 6]
```diff
@@ -172,6 +172,20 @@ Call *method* on the SOAP service.
...
+
+``` javascript
+  client.MyFunction({name: 'value'}).then(function({result, envelope, soapHeader}){
+    // ...
+  }, function(err) {
+    // ...
+  });
+
+  // in async/await flavor
+  const {result, envelope, soapHeader} = await client.MyFunction({name: 'value'});
```
Comment (b-admike) [inline README.md]: no `try/catch` here for the err?

**Pull request merged**